### PR TITLE
Feature/cmaf ham

### PR DESCRIPTION
**Pull request by nicolaslevy**
## Description
This is the CMAF-Ham PoC done in the MonteVIDEO Summer camp 2024. 

HLS and DASH stand as the predominant video streaming technologies currently. Consequently, users often encounter challenges such as converting between HLS and DASH, manipulating manifests, and programmatically understanding manifest structures.

The Common Media Application Format (CMAF) for segmented media (ISO/IEC 23000-19) addresses these challenges by defining a universal format based on ISOBMFF. Additionally, it introduces the Hypothetical Application Model, a framework illustrating the practical usage of CMAF segments and fragments in streaming applications. This project is inspired by the principles outlined in the CMAF standard and the [Hypothetical Application Model] ([Hypothetical Application Model](https://cdn.cta.tech/cta/media/media/resources/standards/cta-5005-a-final.pdf)). 

In this version it is possible to do a basic conversion between HLS <- HAM -> Dash.

Also a new `sample/` folder is added to the project to showcase basic usage of the tools. The `cmaf-ham-converter` is a basic example of how to use CMAF-Ham to convert  HLS <- HAM -> Dash from some basic HLS and DASH Manifests.

## Requirements Checklist
- [ ] Unit Tests updated or fixed
- [ ] Docs/guides updated

### CODE REVIEW

**Comment by nicolaslevy** [comment]
@littlespex During the development of the CMAF-Ham PoC, the Prettier tool was added to format the code. Are you interested in keeping it, or should we remove it?

File: .eslintrc.json
Comment (littlespex) [inline .eslintrc.json:1]: Please revert these changes.

File: .prettierignore
Comment (littlespex) [inline .prettierignore:1]: Please remove prettier

File: .prettierrc
Comment (littlespex) [inline .prettierrc:1]: Please remove prettier

File: package.json [hunk 1]
```diff
@@ -28,6 +28,9 @@
 	"scripts": {
 		"format": "npm run lint -- --fix",
 		"lint": "eslint . --ext .js,.ts",
+		"prettier": "prettier",
+		"prettier:check": "prettier . --ext .js,.ts --check",
+		"prettier:fix": "npm run prettier -- --write",
```
Comment (littlespex) [inline package.json]: remove prettier

File: lib/package.json [hunk 2]
```diff
@@ -68,6 +68,7 @@
 		"build:esm": "node --no-warnings --loader ts-node/esm ../scripts/build.ts",
 		"build:cjs": "tsc --module commonjs --outDir dist/cjs",
 		"test": "node --no-warnings --loader ts-node/esm --test ./test/**/*.test.ts",
+		"test:unit": "node --no-warnings --loader ts-node/esm --test ./src/**/test/*.test.ts",
```
Comment (littlespex) [inline lib/package.json]: All tests should live in `/test/` at the root of the project, so this script should be removed.

File: lib/src/cmaf-ham.ts [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+	AudioTrack,
+	TextTrack,
+	Segment,
+} from './cmaf/ham/types/model/index.js';
+export type * from './cmaf/ham/types/index.js';
+export * from './cmaf/ham/services/getTracks.js';
+export * from './cmaf/ham/services/validate.js';
+export * from './cmaf/ham/services/converters/dashConverter.js';
+export * from './cmaf/ham/services/converters/hlsConverter.js';
+export * as default from './cmaf/ham/index.js';
```
Comment (littlespex) [inline lib/src/cmaf-ham.ts]: No default exports

File: lib/src/cmaf/ham/index.ts
Comment (littlespex) [inline lib/src/cmaf/ham/index.ts:1]: Avoid index files in subfolders. Only the one `index.ts` at the root of the project is needed.

File: lib/src/cmaf/ham/types/index.ts
Comment (littlespex) [inline lib/src/cmaf/ham/types/index.ts:1]: Remove index file. Use explicit named imports.

File: lib/src/cmaf/ham/types/model/index.ts
Comment (littlespex) [inline lib/src/cmaf/ham/types/model/index.ts:1]: Remove index file. Use explicit named imports.

File: lib/src/cmaf/ham/types/model/Track.ts
Comment (littlespex) [inline lib/src/cmaf/ham/types/model/Track.ts:1]: Each type should be in its own file.

File: package.json [hunk 4]
```diff
@@ -46,16 +49,26 @@
...
 		"tslib": "2.6.2",
 		"typedoc": "0.25.8",
 		"typescript": "5.3.3"
+	},
+	"dependencies": {
+		"@types/xml2js": "^0.4.14",
+		"eslint-config-prettier": "^9.1.0",
+		"eslint-plugin-prettier": "^5.1.3",
+		"m3u8-parser": "^7.1.0",
+		"prettier": "^3.2.5",
```
Comment (littlespex) [inline package.json]: Remove all prettier dependencies.

File: .nvmrc
Comment (littlespex) [inline .nvmrc:1]: Please remove this file

File: lib/package.json [hunk 5]
```diff
@@ -68,6 +68,7 @@
...
 		"prepublishOnly": "npm run build"
 	}
-}
+}
```
Comment (littlespex) [inline lib/package.json]:
```suggestion
}

```

File: lib/src/cmaf-ham.ts [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+	Presentation,
+	SelectionSet,
+	AlignedSwitchingSet,
+	SwitchingSet,
+	VideoTrack,
+	AudioTrack,
+	TextTrack,
+	Segment,
+} from './cmaf/ham/types/model/index.js';
+export type * from './cmaf/ham/types/index.js';
```
Comment (littlespex) [inline lib/src/cmaf-ham.ts]: Remove exports from sub folder index files in favor of direct exports from the named source files.

File: lib/src/cmaf/ham/mapper/DashMapper.ts [hunk 7]
```diff
@@ -0,0 +1,33 @@
+import { mapDashToHam } from './dash/mapDashToHam.js';
+import { xmlToJson } from '../utils/xmlUtils.js';
+import { mapHamToDash } from './dash/mapHamToDash.js';
+import { addMetadataToDash, getMetadata } from '../utils/manifestUtils.js';
+import { IMapper } from './IMapper.js';
+import type { DashManifest, Manifest } from '../types';
```
Comment (littlespex) [inline lib/src/cmaf/ham/mapper/DashMapper.ts]: Import these directly from the named files. (All imports should be from `.js` files)

File: lib/src/cmaf/ham/mapper/HlsMapper.ts [hunk 8]
```diff
@@ -0,0 +1,24 @@
+import type { Manifest } from '../types';
```
Comment (littlespex) [inline lib/src/cmaf/ham/mapper/HlsMapper.ts]: Import these directly from the named files. (All imports should be from .js files)

File: lib/src/cmaf/ham/mapper/IMapper.ts [hunk 9]
```diff
@@ -0,0 +1,10 @@
+import type { Presentation } from '../types/model/index.js';
+import type { Manifest } from '../types';
```
Comment (littlespex) [inline lib/src/cmaf/ham/mapper/IMapper.ts]: Import these directly from the named files. (All imports should be from .js files)

File: lib/src/cmaf/ham/mapper/MapperContext.ts [hunk 10]
```diff
@@ -0,0 +1,34 @@
+import { IMapper } from './IMapper.js';
+import type { Manifest } from '../types';
```
Comment (littlespex) [inline lib/src/cmaf/ham/mapper/MapperContext.ts]: Import these directly from the named files. (All imports should be from .js files)

File: lib/src/cmaf/ham/utils/manifestUtils.ts [hunk 11]
```diff
@@ -0,0 +1,40 @@
+import type { DashManifest, Manifest } from '../types';
```
Comment (littlespex) [inline lib/src/cmaf/ham/utils/manifestUtils.ts]: Import from name source files.

File: lib/src/cmaf/ham/utils/manifestUtils.ts
Comment (littlespex) [inline lib/src/cmaf/ham/utils/manifestUtils.ts:1]: Move each utility function into its own file.

File: lib/src/cmaf/ham/utils/manifestUtils.ts [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+
+export function getMetadata(manifest: Manifest | undefined): JSON | undefined {
```
Comment (littlespex) [inline lib/src/cmaf/ham/utils/manifestUtils.ts]: `JSON` isn't a valid return type in this context. The results of calling `JSON.parse` would be an object of `any` type.

File: lib/src/cmaf/ham/utils/utils.ts
Comment (littlespex) [inline lib/src/cmaf/ham/utils/utils.ts:1]: Move these utility function into their own files.

File: lib/src/cmaf/ham/mapper/HlsMapper.ts [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+import type { Presentation } from '../types/model/index.js';
+import { IMapper } from './IMapper.js';
+import { mapHamToHls } from './hls/mapHamToHls.js';
+import { mapHlsToHam } from './hls/mapHlsToHam.js';
+import { getMetadata } from '../utils/manifestUtils.js';
+
+export class HlsMapper implements IMapper {
+	private manifest: Manifest | undefined;
+
+	getManifestMetadata(): JSON | undefined {
```
Comment (littlespex) [inline lib/src/cmaf/ham/mapper/HlsMapper.ts]: `JSON` isn't a valid type in this context. This should be typed as `any`.

File: lib/src/cmaf/ham/mapper/IMapper.ts [hunk 14]
```diff
@@ -0,0 +1,10 @@
...
+
+export interface IMapper {
+	toHam(manifest: Manifest): Presentation[];
+
+	toManifest(presentation: Presentation[]): Manifest;
+
+	getManifestMetadata(): JSON | undefined;
```
Comment (littlespex) [inline lib/src/cmaf/ham/mapper/IMapper.ts]: `JSON` isn't a valid type in this context. It refers to the window level `JSON` object that contains `JSON.stringify`, `JSON.parse`, etc. The result of calling `JSON.parse` is an object with an `any` type.

File: lib/src/cmaf/ham/mapper/dash/mapDashToHam.ts
Comment (littlespex) [inline lib/src/cmaf/ham/mapper/dash/mapDashToHam.ts:1]: Move each function into it's own file.

File: package.json [hunk 15]
```diff
@@ -46,16 +49,26 @@
...
 		"ts-node": "10.9.2",
 		"tslib": "2.6.2",
 		"typedoc": "0.25.8",
 		"typescript": "5.3.3"
+	},
+	"dependencies": {
+		"@types/xml2js": "^0.4.14",
+		"eslint-config-prettier": "^9.1.0",
+		"eslint-plugin-prettier": "^5.1.3",
+		"m3u8-parser": "^7.1.0",
```
Comment (littlespex) [inline package.json:66]: I understand that some of the new utility functions require these to be non-dev dependencies, but this means that someone using the library for CMCD, or ID3, has to install these packages. I'm not sure what the best solution for this is, `optionalDependecies` perhaps, but I think we might move towards breaking all of the features out into their own npm packages: `@svta/cml-cmcd`, `@svta/cml-id3`, `@svta/cml-cmaf-ham`. Nothing actionable here, just mentioning it for future reference.

File: package.json [hunk 16]
```diff
@@ -46,16 +49,26 @@
 	],
 	"devDependencies": {
 		"@microsoft/api-extractor": "7.40.1",
+		"@types/fs-extra": "11.0.3",
 		"@types/node": "20.11.17",
 		"@typescript-eslint/eslint-plugin": "7.0.1",
 		"@typescript-eslint/parser": "7.0.1",
 		"eslint": "8.56.0",
 		"eslint-plugin-tsdoc": "0.2.17",
+		"fs-extra": "11.1.1",
```
Comment (littlespex) [inline package.json]: I don't see `fs-extra` being used anywhere, so these dependencies should be removed.

File: samples/cmaf/ham/cmaf-ham-conversion/.gitignore
Comment (littlespex) [inline samples/cmaf/ham/cmaf-ham-conversion/.gitignore:1]: Remove this file in favor of the root `.gitignore`. `node_modules` is already ignored at the root level, and `output` should be renamed to `dist` to match the other places in the repo where compiled/generated files are created.

File: samples/cmaf/ham/cmaf-ham-conversion/package.json [hunk 17]
```diff
@@ -0,0 +1,23 @@
...
+    "version": "1.0.0",
+    "description": "",
+    "main": "index.js",
+    "type": "module",
+    "scripts": {
+        "dev": "node --loader ts-node/esm src/index.ts",
+        "serve": "serve --cors ./output"
+    },
+    "dependencies": {
+        "@svta/common-media-library": "file:../../../../../../lib/",
```
Comment (littlespex) [inline samples/cmaf/ham/cmaf-ham-conversion/package.json]: In a monorepo, this should be be:
```suggestion
        "@svta/common-media-library": "*",
```

File: samples/cmaf/ham/cmaf-ham-conversion/package.json [hunk 18]
```diff
@@ -0,0 +1,23 @@
...
+        "fs": "^0.0.1-security",
+        "ts-node": "^10.9.2",
+        "tsc": "^2.0.4"
```
Comment (littlespex) [inline samples/cmaf/ham/cmaf-ham-conversion/package.json]: These dependencies aren't needed as they are already installed in the root package file.

File: samples/cmaf/ham/cmaf-ham-conversion/package.json
Comment (littlespex) [inline samples/cmaf/ham/cmaf-ham-conversion/package.json:1]: For now, lets remove the sub folders `cmaf/ham/` and just have `cmaf-ham-conversion` at the root of the `samples` folder.

File: samples/cmaf/ham/cmaf-ham-conversion/package-lock.json
Comment (littlespex) [inline samples/cmaf/ham/cmaf-ham-conversion/package-lock.json:1]: Remove this package lock file. This folder should be added to the root `package.json`'s `workspace` array, and all dependencies should be installed from the root of the monorepo using the workspace flag. For example:
```sh
npm i serve -w samples/cmaf/ham/cmaf-ham-conversion/ -D -E
```

File: package.json
Comment (littlespex) [inline package.json:1]: Add `samples/*` to the `workspaces` array.

**Review by littlespex — CHANGES_REQUESTED** [review]
I've looked over the PR at a high level and there are a number of individual comments, but they all fall into a few basic categories, and they all relate to matching the existing conventions used in the rest of the project:
- One named export per file (and no default exports).
- No `index.ts` files, except the one at the root of the project.
- All imports should be from the named source file, not indexes, and should contain `.js` extensions.
- Remove all unused dependencies. This includes prettier. If anyone wants to discuss changing listing/formatting tools, open a new ticket.
- Remove unnecessary files, like `.nvmrc` and `samples/.gitignore`
- Use npm's workspaces for the new `samples` package

File: lib/src/cmaf/ham/utils/manifestUtils.ts (hunk 12 above)
Comment (nicolaslevy) [reply lib/src/cmaf/ham/utils/manifestUtils.ts]: Ok, if you agree for now let's return 'any'. This functions will change in the future as we should re-think how store and convert data like "CMCDParameters" from the manfiests.

Comment (littlespex) [reply lib/src/cmaf/ham/utils/manifestUtils.ts]: Agreed, `any` will work for now.

**Comment by nicolaslevy** [comment]
@littlespex I made all the changes, please review this PR again :)

File: lib/src/cmaf/ham/types/mapper/dash/AdaptationSet.ts [hunk 19]
```diff
@@ -0,0 +1,42 @@
...
+import type { SegmentTemplate } from './SegmentTemplate.js';
+
+/**
+ * DASH Adaptation Set
+ *
+ * @group CMAF
+ * @alpha
+ */
+export type AdaptationSet = {
+	$: {
```
Comment (littlespex) [inline lib/src/cmaf/ham/types/mapper/dash/AdaptationSet.ts:15]: Why do all of these types have a `$` property? I assume it's the XML node's attributes list, but I'd rather not have users make assumptions about the code.

Comment (nicolaslevy) [reply lib/src/cmaf/ham/types/mapper/dash/AdaptationSet.ts:15]: Yes, your assumption is correct according to the xml2js documentation (https://www.npmjs.com/package/xml2js). The $ object in the output from xml2js  represents the attributes of the XML node. For example, in XML like:

```
<person name="John" age="30">
  <phone>123456</phone>
</person>
```

xml2js would convert it into a JavaScript object where the $ object contains the attributes of the \<person\> node

```
{
  person: {
    $: { name: "John", age: "30" },
    phone: "123456"
  }
}
```

In this implementation, DashManifest (and all its sub-parts) is tightly coupled with xml2js. Perhaps we could leave a comment or something to that effect. What do you think @littlespex ?

Comment (littlespex) [reply lib/src/cmaf/ham/types/mapper/dash/AdaptationSet.ts:15]: The tight coupling is something we will need to address at some point. I know dash.js specifically moved away from xml2json, and we want CML to be as dependency free as possible. We can leave this in place for now while we come up with a better solution.

**Review by littlespex — APPROVED** [review]
I committed a few semantic changes, separating types into their own files. I'm not sure if this will cause any issues with DCO once this branch is merged upstream, but I will take care of them if they do occur.

File: lib/src/cmaf/ham/types/mapper/dash/AdaptationSet.ts (hunk 19 above)
Comment (nicolaslevy) [reply lib/src/cmaf/ham/types/mapper/dash/AdaptationSet.ts:15]: Agree, I think we should think about using stuff from dash.js to parse the Dash Manifest and maybe that could be a contribution for dash.js also

**Pull request merged**